### PR TITLE
Implement Additional Quantized Utils

### PR DESCRIPTION
**Pull request by ShikharJ**
Major overhaul of Arun's existing PR on additional quantized routines. This needs to be modified post the merging of the MBConv PR.

### CODE REVIEW

**Comment by ShikharJ** [comment]
@harsha-simhadri This is now ready for reviewing. MaxPool and Convolution operators would be pushed in a separate PR.

File: c_reference/include/quantized_utils.h [hunk 1]
```diff
@@ -53,15 +53,261 @@ void v_q_tanh(const INT_T* const vec, ITER_T len, INT_T* const ret,
...
+ * @param[in]       scret     scaling factor for output vector
+ * @param[out]      ret       pointer to the output vector
+ * @return          none
+ * @example         vec       = {13, 54, 34, 35, 87}
+ *                  len       = 5
+ *                  scvec     = 8
+ *                  scret     = 8
+ *                  ret       = {40, 6832, 560, 635, 29493}
+ */
+void v_q_exp(const INT_T* const vec, ITER_T len, INT_T* const ret,
```
Comment (harsha-simhadri) [inline c_reference/include/quantized_utils.h:101]: does it saturate?

File: c_reference/include/quantized_utils.h [hunk 2]
```diff
@@ -53,15 +53,261 @@ void v_q_tanh(const INT_T* const vec, ITER_T len, INT_T* const ret,
...
+ *                                {4, 5, 6} }
+ *                  ret       = { {1,  4},
+ *                                {2,  5},
+ *                                {3,  6} }
+ */
+void m_q_transpose(const INT_T* const mat, ITER_T nrows, ITER_T ncols,
+                   INT_T* const ret);
+/**
+ * @brief Performs the row-order or the column-order reversal of the input matrix.
+ * @param[in]       mat       pointer to the matrix on which reversal is to be performed
```
Comment (harsha-simhadri) [inline c_reference/include/quantized_utils.h:150]: row major or col major?

File: c_reference/include/quantized_utils.h [hunk 3]
```diff
@@ -53,15 +53,261 @@ void v_q_tanh(const INT_T* const vec, ITER_T len, INT_T* const ret,
...
+ * @example         mat       = { {1, 2, 3},
+ *                                {4, 5, 6} }
+ *                  ret       = { {1,  4},
+ *                                {2,  5},
+ *                                {3,  6} }
+ */
+void m_q_transpose(const INT_T* const mat, ITER_T nrows, ITER_T ncols,
+                   INT_T* const ret);
+/**
+ * @brief Performs the row-order or the column-order reversal of the input matrix.
```
Comment (harsha-simhadri) [inline c_reference/include/quantized_utils.h:149]: how general is this function?

File: c_reference/include/quantized_utils.h [hunk 4]
```diff
@@ -53,15 +53,261 @@ void v_q_tanh(const INT_T* const vec, ITER_T len, INT_T* const ret,
...
 // Function for multiplying a matrix with a vector.
 void m_q_mulvec(const INT_T* const mat, const INT_T* const vec, ITER_T nrows,
                 ITER_T ncols, INT_T* const ret, SCALE_T scmat, SCALE_T scvec,
                 SCALE_T H1, SCALE_T H2);
+/**
+ * @brief Performs sparse matrix multiplication of a matrix and a vector.
+ * mat_indices and mat_values combined are a sparse representation; dim(mat_values) = [ndims], dim(mat_indices) = [ndims + ncols].
+ * mat_values[i] is the i^th non-zero value of the input matrix, and mat_indices[i] encodes the location of mat_values[i].
+ * Number of zeroes before Aidx[i] : row of Aval[i]
+ * Aidx[i] + ... + Aidx[l] where l is the largest value less than i such that A[idx] = 0 : column of Aval[i]
```
Comment (harsha-simhadri) [inline c_reference/include/quantized_utils.h:223]: not sure I understand the layout. Is this CSR? where are matrix dimensions in this interface?

File: c_reference/src/quantized_utils.c [hunk 5]
```diff
@@ -139,22 +150,185 @@ void v_q_scalar_mul(INT_T scalar, const INT_T* const vec, ITER_T len,
...
+      vec[i] /= scvec;
+    #endif
+  }
+}
+
+void m_q_transpose(const INT_T* const mat, ITER_T nrows, ITER_T ncols,
+                   INT_T* const ret) {
+  for (ITER_T i = 0; i < nrows; i++) {
+    for (ITER_T j = 0; j < ncols; j++) {
+      ret[i * ncols + j] = mat[j * nrows + i];
```
Comment (harsha-simhadri) [inline c_reference/src/quantized_utils.c:206]: *dst++ = *src;
src+= stride;

File: c_reference/src/quantized_utils.c [hunk 6]
```diff
@@ -139,22 +150,185 @@ void v_q_scalar_mul(INT_T scalar, const INT_T* const vec, ITER_T len,
...
+    v_q_treesum(&treesumBuffer[0], ncols, H1, H2);
+    #ifdef SHIFT
+      ret[row] = (treesumBuffer[0] >> (scmat + scvec));
+    #else
+      ret[row] = ((treesumBuffer[0] / scmat) / scvec);
+    #endif
+  }
+}
+
+void m_q_sparse_mulvec(const ITER_T* const mat_indices, const INT_T* const mat_values,
```
Comment (harsha-simhadri) [inline c_reference/src/quantized_utils.c:278]: col_indices instead of mat_indices

File: c_reference/include/quantized_utils.h (hunk 1 above)
Comment (ShikharJ) [reply c_reference/include/quantized_utils.h:104]: Added a comment regarding the caveat.

**Comment by ShikharJ** [comment]
@harsha-simhadri Let's get this one merged first, I'd reckon there might be some merge conflicts with the other PR.

**Pull request merged**